src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -83,22 +83,95 @@ public Dictionary(int capacity, IEqualityComparer<TKey>? comparer)
         public Dictionary(IDictionary<TKey, TValue> dictionary) : this(dictionary, null) { }
 
         public Dictionary(IDictionary<TKey, TValue> dictionary, IEqualityComparer<TKey>? comparer) :
-            this(dictionary != null ? dictionary.Count : 0, comparer)
+            this(0, comparer)
         {
             if (dictionary == null)
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.dictionary);
             }
 
+            int count = dictionary.Count;
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
+            }
+            if (count == 0)
+            {
+                // Nothing to copy, all done
+                return;
+            }
+
             // It is likely that the passed-in dictionary is Dictionary<TKey,TValue>. When this is the case,
             // avoid the enumerator allocation and overhead by looping through the entries array directly.
             // We only do this when dictionary is Dictionary<TKey,TValue> and not a subclass, to maintain
             // back-compat with subclasses that may have overridden the enumerator behavior.
             if (dictionary.GetType() == typeof(Dictionary<TKey, TValue>))
             {
                 Dictionary<TKey, TValue> d = (Dictionary<TKey, TValue>)dictionary;
-                int count = d._count;
+
+                Debug.Assert(d._buckets is not null);
+                Debug.Assert(d._entries is not null);
+
+                Entry[] oldEntries = d._entries;
+                if (d._comparer == _comparer)
+                {
+                    // Comparers are the same, can copy dictionary wholesale
+                    int size = HashHelpers.GetPrime(count);
+                    if (size == oldEntries.Length)
```

Comment:
With this condition, can the arrays be allocated using `this(dictionary != null ? dictionary.Count : 0, comparer)` like before?

Reply:
Yes I think so